packages/interface/src/components/code-editor/index.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import Textarea from 'react-autosize-textarea';
+
+/**
+ * WordPress dependencies
+ */
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useState } from '@wordpress/element';
+import { useInstanceId } from '@wordpress/compose';
+import { VisuallyHidden } from '@wordpress/components';
+
+export default function CodeEditor( { value, onChange, onInput } ) {
```

Comment:
Should this be more a `@wordpress/components` component instead?

Reply:
The component does not seem very generic and the styles are a little bit specific to our needs (I think it would only be useful for someone building an editor behaving exactly like ours).  But I can move some styles to CodeEditorScreen and try to make it more generic.
Another option is to not have a CodeEditor component and expose it under CodeEditorScreen something like CodeEditorScreen.Editor.
Let me know what you would prefer.

Reply:
It bothers me a bit to have "Code Editor" in "@interface" package with or without screen. Interface should be about building screens why do we need to know about code editor.

Reply:
Would it make sense to introduce a new `@wordpress/code-editor` package (akin to `@wordpress/editor`)?

Reply:
> Would it make sense to introduce a new @wordpress/code-editor package (akin to @wordpress/editor)?

I think that makes sense but because I thought it'd be smallish I thought we could just move the component here to the components package.

Reply:
> I think that makes sense but because I thought it'd be smallish

I agree, the package although making sense conceptually would be very small.

> I thought we could just move the component here to the components package.

I can make a try to make it part of the components package but it seems very specific with our dimensions, our styles, etc.. This will not be useful to anyone outside us and having these components as part of our public API just makes it something we need to keep forever. For example, if we had not exposed PostTextEditor on wp.editor Probably right now I would not divide between CodeEditorScreen and CodeEditor and would just have a single component. But as we need to keep PostTextEditor I'm having a CodeEditor component so PostTextEditor can just use it. I think adding more components to wp.components will just be something more we have to maintain in the future.
Ideally, we just want to share the code between edit-post and edit-site. The component should look mostly equal in both places and I think repeating hundreds of lines of code that we need to change in two places each time the UI changes is not a good idea. The interface package was a simple solution it is private can be used by edit-post and edit site.

We may consider creating another private package to put artifacts reused between editor, edit-post, and edit-site that we don't think are something worth exposing (maybe named editor-utils?). The screens are mostly equal and we should make it easy to share artifacts between them without having to think much.

If we don't create a new package:

If we add CodeEditor to wp.components can we keep CodeEditorScreen on wp.interface? Or would you prefer if I change it to SimpleExitableScreen and labels referencing code editor would be passed as props? If we don't keep CodeEditorScreen on the Interface package what would be the ideal place to keep it?


Regarding CodeEditor, another alternative instead of exposing is to change wp.editor.PostTextEditor that we already have to also accept value, onChange, onInput. When the props are not passed it would behave exactly as it does now and retrieve them assuming what is being edited is the current post if the props are passed it works in a generic way.
Do we prefer an update to wp.editor.PostTextEditor or expose a new component under wp.components?

Reply:
Gotcha, also makes sense. The separate package could maybe make sense if we feel that the code editor is too specific to live in the general-purpose `@wordpress/components` package, as Jorge suggested.

Reply:
> The component should look mostly equal in both places and I think repeating hundreds of lines of code that we need to change in two places each time the UI changes is not a good idea. 

When did the last time the UI changed? 

> The interface package was a simple solution it is private can be used by edit-post and edit site.

We had this issue for all the UI of the screens so far, and every time we try to share code, personally, it feels wrong to me because we just share code to share code and we don't think about the right abstractions/packages and cripple them.

--
I understand that it's just easy for developers and don't repeat yourself and and all of that but I can't stop but feel that it's wrong. We don't make the necessary efforts to build the right abstraction and we just want to ship our code. For me if the goal is to just ship the feature, it's better to duplicate. If we want to share we should start thinking about the right abstraction, the right packages.

I want to use the interface package today, it became so crippled that I won't ever look for it because it's not clear what is it for, it just there for convenience which is sad. I guess we can continue that trend because it's already a very bad package. I won't block this anymore.

Reply:
A long time ago, I proposed that editSite is just editPost with another function to initialize it. I think that would have been the best solution but probably too late now.

> We may consider creating another private package to put artifacts reused between editor, edit-post, and edit-site that we don't think are something worth exposing (maybe named editor-utils?). 

I think that's what "interface" has become, we can either continue the trend or start over and try to clean "interface" and add a new package because make it clear that it's just to share code between edit-post and edit-site and ideally shouldn't exist.

Reply:
Hi @youknowriad, I did not have a strong preference for extending the interface package. I updated the PR to not add anything to the interface package and follow the duplication approach.

> A long time ago, I proposed that editSite is just editPost with another function to initialize it. I think that would have been the best solution but probably too late now.

This would have been the right approach. Sharing functionality between edit-site and edit-post is always hard while their UI is mostly the same. But I guess now we can not make them a single package.
The interface package already says "Interface module for WordPress. The package contains shared functionality across the modern JavaScript-based WordPress screens." so the idea of the package is already to share functionality between the screens and the package exists just for that but we still have a higher standard of what should be an abstraction there vs what is a component when only internal to a package and the current components there are more generic than what we had here and contain readme files etc.